### PR TITLE
Packet loopback action for tunnels

### PR DESCRIPTION
**Pull request by JaiOCP**
Added loopback action as switch attribute: This will apply to all the tunnels configured
Added loopback action as tunnel attribute: This will apply as a tunnel specific action.

### CODE REVIEW

**Review by kcudnik — APPROVED** [review]

File: inc/saitunnel.h [hunk 1]
```diff
@@ -643,6 +643,15 @@ typedef enum _sai_tunnel_attr_t
      */
     SAI_TUNNEL_ATTR_TERM_TABLE_ENTRY_LIST,
 
+    /**
+     * @brief Packet action when a packet ingress and gets routed back to same tunnel
+     *
+     * @type sai_packet_action_t
+     * @flags CREATE_AND_SET
+     * @default SAI_PACKET_ACTION_FORWARD
```
Comment (prsunny) [inline inc/saitunnel.h:651]: should the default be SAI_PACKET_ACTION_DROP?

Comment (JaiOCP) [reply inc/saitunnel.h:651]: Most of the usecases I am familiar with are forward for eg MPLS, IPinIP and GRE with L3 payload.

**Comment by rlhui** [comment]
What if the next-hop after decap is another tunnel, but not this same tunnel where pkt is received from?  Say if the next-hop for the dst is an ecmp?

**Comment by rlhui** [comment]
Can we increment a drop reason counter when it's dropped due to this?

**Comment by JaiOCP** [comment]
> What if the next-hop after decap is another tunnel, but not this same tunnel where pkt is received from? Say if the next-hop for the dst is an ecmp?

If packet is forwarded to some other tunnel instead of the ingress tunnel then this attribute will not be applied. Only when packet loops back into the ingress tunnel then this attribute is applicable.

**Comment by JaiOCP** [comment]
> Can we increment a drop reason counter when it's dropped due to this?

yes, I am thinking to add switch level attribute: SAI_SWITCH_STAT_TUNNEL_LOOPBACK_PACKET_DROP
@itaibaz Is this OK?

**Comment by itaibaz** [comment]
Hi, let me check internally and update in several days, thanks

**Comment by JaiOCP** [comment]
> We prefer using
> saidebugcounter.h:sai_out_drop_reason_t:SAI_OUT_DROP_REASON_TUNNEL_LOOPBACK_PACKET_DROP
> As this is already existing mechanism used to debug drops, and in specific tunnel drops, in sonic

Done

**Comment by JaiOCP** [comment]
@rlhui @itaibaz @prsunny please review and approve

**Review by itaibaz — APPROVED** [review]

**Review by rlhui — APPROVED** [review]

**Pull request merged**